thavalon-server/src/lobby/mod.rs

```diff
@@ -0,0 +1,270 @@
+//! Module for all game lobby code. The Lobby represents the interface between
+//! the actual game, the database, and player connections.
+
+use crate::database::games::{DBGameError, DBGameStatus, DatabaseGame};
+use crate::game::{builder::GameBuilder, Action, Message};
+
+use futures::{FutureExt, SinkExt, StreamExt};
+use thiserror::Error;
+use tokio::{
+    sync::{
+        mpsc::{self, Receiver, Sender},
+        oneshot, Mutex,
+    },
+    task,
+};
+use warp::filters::ws::{self, WebSocket};
+
+use std::collections::HashMap;
+use std::sync::Arc;
+
+/// Type repreesenting a channel to the lobby to issue commands.
+pub type LobbyChannel = Sender<(LobbyCommand, ResponseChannel)>;
+
+/// Type representing a oneshot sender back to the caller.
+pub type ResponseChannel = oneshot::Sender<LobbyResponse>;
+
+/// Enum of possible lobby-related errors.
+#[derive(Debug, Error)]
+pub enum LobbyError {
+    #[error("An error occurred while updating the database.")]
+    DatabaseError,
+    #[error("The player is already in the game.")]
+    DuplicatePlayerError,
+    #[error("A lobby update was attemped in a state that doesn't permit this update.")]
+    InvalidStateError,
+    #[error("An unknown error occurred. See logs for details.")]
+    UnknownError,
+}
+
+/// Enum of available commands to send to the lobby.
+pub enum LobbyCommand {
+    AddPlayer {
+        player_id: String,
+        display_name: String,
+    },
+    GetFriendCode,
+    IsPlayerRegistered {
+        player_id: String,
+    },
+    ConnectClientChannels {
+        player_id: String,
+        ws: WebSocket,
+    },
+}
+
+/// Enum of possible responses from the lobby.
+#[derive(Debug)]
+pub enum LobbyResponse {
+    Standard(Result<(), LobbyError>),
+    FriendCode(String),
+    IsPlayerRegistered(bool),
+}
+
+/// Represents connections to and from the game for an individual player.
+#[derive(Debug)]
+pub struct PlayerClient {
+    to_game: Sender<Action>,
+    from_game: Arc<Mutex<Receiver<Message>>>,
+}
+
+/// A lobby for an individual game. The Lobby acts as an interface between the
+/// Thavalon game instance, the DatabaseGame which keeps the game state in sync
+/// with the database, and all players connected to the game.
+pub struct Lobby {
+    database_game: DatabaseGame,
+    friend_code: String,
+    players: HashMap<String, PlayerClient>,
+    status: DBGameStatus,
```

Comment:
Why does `Lobby` have status and friend code fields on top of the ones in `DatabaseGame`?

Reply:
It was mainly for convenience. I've left them for now as this is still a WIP, but I may get rid of them or decouple them from the DB in the future.